[Colex2000] iP

Colex2000 opened the pull request:
## Demon
> You only live once

The demon chat bot frees your brain memory space as it helps you remember the tasks you have to do. It's 
* **FREE** & EASY to use
* ~~FAST~~ NOT JUST FAST, SUPER FAST!
* Easy to learn
* Lastly, ~~washes your plates~~. KIDDING 💛

All you need to do is:

1. Download it from [here](https://github.com/Colex2000/ip)
2. Start the bot
3. Add tasks by following the commands
4. BooM! Demon will manage it for you!

Features include:
- [x] Adding tasks 
- [x] Deleting tasks
- [x] Mark/Unmark tasks as done/not done
- [x] List tasks

and more coming soon! 

___

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(Demon.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Colex2000 committed:
Changed chatbot name from "Duke" to "Demon". Added new lines to greet and exit.

Colex2000 committed:
Changed file name to 'Demon'.
Added echo from the Demon and exit.

Colex2000 committed:
Added function to add to-do to list.
Added function to display to-do list.

Colex2000 committed:
Added Task.java file.
Added the ability to mark tasks as done and undone.

Colex2000 committed:
Added ToDo, Deadline, Event class file.
Modified Deadline class.

Colex2000 committed:
Added EmptyDescriptionException and NoTimingException exception class file.
Handled exceptions across all commands.

Colex2000 committed:
Added delete command.

Colex2000 committed:
Edited runtest.bat file

Colex2000 committed:
Revert "Edited runtest.bat file"

This reverts commit 3ee67e23851ba57e0c7ddd29be5bf037f71e388b.

Colex2000 committed:
Edited runtest.bat file

Colex2000 committed:
Edited runtest.bat file.

Colex2000 committed:
Merge remote-tracking branch 'origin/master'

Colex2000 committed:
Delete ip/text-ui-test directory

Colex2000 committed:
Add task list save to and load from file

Colex2000 committed:
Merge branch 'branch-Level-7'

Colex2000 committed:
Add re-formatting of date and time

Colex2000 committed:
Merge branch 'branch-Level-8'

Colex2000 committed:
Update files and logics according to OOP principles

Colex2000 committed:
Merge branch 'A-MoreOOP'

Colex2000 committed:
Refactor: Consolidate all classes into a single package

Colex2000 committed:
Refactor: Deadline class into demon package

Colex2000 committed:
Merge branch 'A-Packages'

Colex2000 committed:
Merge branch 'add-gradle-support'

Colex2000 committed:
Merge branch 'add-gradle-support'

Colex2000 committed:
Add tests for Deadline string representation Storage load method

Colex2000 committed:
Merge branch 'A-JUnit'

Colex2000 committed:
Add function to find task by searching for a keyword.

Colex2000 committed:
Update code formatting to align with coding style standards

Colex2000 committed:
Add Javadoc headers for better API documentation

Colex2000 committed:
Add MANIFEST.MF

Colex2000 committed:
Merge branch 'A-CodingStandard'

Colex2000 committed:
Merge branch 'A-JavaDoc'

Colex2000 committed:
Merge branch 'Level-9'

# Conflicts:
#	src/main/java/demon/Commands.java

Colex2000 committed:
Update file according to renamed bot "Duke" to "Demon"

Colex2000 committed:
Merge branch 'add-gradle-support'

Colex2000 committed:
Merge commit '2e7e7e2854541e415e26c76edc830081d8b2ef09'

ziyang27 commented on src/main/java/demon/Commands.java:
```diff
@@ -0,0 +1,285 @@
+ * based on user input. Commands such as 'list', 'unmark', 'mark', 'delete', and others are supported, each
+ * corresponding to a specific method within this class.
+ */
+public class Commands {
```
Hi Cole! Small point here, but maybe you can consider naming it just 'Command' as the plural form make it seem like it is a collection of commands? 🤔

drustanyjt replied:
Second this!

ziyang27 commented on src/main/java/demon/Commands.java:
```diff
@@ -0,0 +1,285 @@
+     *
+     * @param tasks an array containing all tasks.
+     */
+    public void list(TaskList tasks) {
```
I like how you used 'tasks' to show that TaskList contains multiple tasks 👍

Colex2000 replied:
Thank you 🙏

ziyang27 commented on src/main/java/demon/Commands.java:
```diff
@@ -0,0 +1,285 @@
+        LocalDateTime dateTime = LocalDateTime.parse(by, formatter);
+        Deadline itemDeadline = new Deadline(description, dateTime);
+        tasks.getTaskList().add(itemDeadline);
+        String stringToSave = "D | " + (itemDeadline.getStatusIcon().equals("X") ? "1" : "0")
+                            + " | " + description + " | " + by + "\n";
```
Hi, can't I really see from here but I think the indentation for the wrapped line is more than 8 spaces 🤔 Small issue, can just take a quick look at your IDE to verify, hehe 😄

ziyang27 commented on src/main/java/demon/Deadline.java:
```diff
@@ -0,0 +1,27 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDateTime dateTime;
```
Hmm, perhaps you can consider changing the accessibility to 'private' and use a getter? 🤔

ziyang27 commented on src/main/java/demon/Demon.java:
```diff
@@ -0,0 +1,128 @@
+package demon;
+
+import java.nio.file.NoSuchFileException;
+
+import java.io.IOException;
+
+import java.util.Scanner;
+
+import java.time.format.DateTimeParseException;
```
Small issue, but I think you accidentally added a new line for every import. Additionally, perhaps you can try ordering them in a specific way (alphabetical) and make it consistent throughout the repo? 🤔

Colex2000 replied:
Sure, thanks!

ziyang27 commented on src/main/java/demon/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * An exception class that will be thrown when description of a task is not provided from the input.
+ */
+public class EmptyDescriptionException extends Exception {
```
Nice exception, but perhaps you can try making a DemonException and inherit all demon-related exceptions from it? 👍

Colex2000 replied:
Good idea, done!

ziyang27 commented on src/main/java/demon/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+ * Facilitates the loading of tasks from taskList.txt file.
+ */
+public class TaskList {
+    private final List<Task> taskList;
```
Perhaps you can name this as 'tasks'? 👍

Colex2000 replied:
Sure!

ziyang27 reviewed (commented):
LGTM! In general, the namings are very clear and the code is clean. There are just some minor suggestions for the author to consider. Keep it up 👍

Colex2000 committed:
Update name of 'taskList' to 'tasks'

Colex2000 committed:
Refactor: Introduce DemonException class for custom exceptions

Colex2000 committed:
Reformat: Delete extra line space

Colex2000 committed:
Refactor: Change class name and variable accessibility

Colex2000 committed:
Reformat: proper indentation for wrapped text

Colex2000 committed:
Add: GUI for Demon chatbot

drustanyjt commented on src/main/java/demon/Commands.java:
```diff
@@ -0,0 +1,285 @@
+            System.out.println(Ui.notWithinRange() + tasks.getTaskList().size());
+        } else {
+            Task item = tasks.getTaskList().get(num-1);
+            if (item.getStatusIcon().equals("X")) {
```
"X" seems to be a magic value that relies on the implementation of Task. It might be easier (and safer) to abstract this out to another boolean method such as `item.isMarked()` that returns true if the item is indeed marked, and false otherwise?

drustanyjt commented on src/main/java/demon/Commands.java:
```diff
@@ -0,0 +1,285 @@
+            System.out.println(Ui.notWithinRange() + tasks.getTaskList().size());
+        } else {
+            Task item = tasks.getTaskList().get(num-1);
+            if (item.getStatusIcon().equals(" ")) {
```
" " seems to be a magic value that relies on the implementation of Task. It might be easier (and safer) to abstract this out to another boolean method such as item.isMarked() that returns true if the item is indeed marked?

drustanyjt commented on src/main/java/demon/Commands.java:
```diff
@@ -0,0 +1,285 @@
+            throw new EmptyDescriptionException("Description is EMPTY!");
+        }
+        String toDo = parts[1];
+        Todo item_toDo = new Todo(toDo);
```
Should 'item_toDo' be in `camelCase` withouth the underscore since it's a variable?

drustanyjt commented on src/main/java/demon/Commands.java:
```diff
@@ -0,0 +1,285 @@
+            throw new EmptyDescriptionException("Description is EMPTY!");
+        }
+        String description_date = parts[1];
+        String[] parts2 = description_date.split("/from ");
```
I feel this might be deviating from the coding style as well since `parts2` is not strictly in `camelCase`

drustanyjt commented on src/main/java/demon/Demon.java:
```diff
@@ -0,0 +1,128 @@
+     */
+    public Demon(String filePath) {
+        ui = new Ui();
+        final Storage STORAGE = new Storage(filePath);
```
@Colex2000 You use `filePath` as both a parameter in your constructor (which you use in your `main` method in line 125, and as a instance variable (in line 17). I think this is an instance of variable shadowing which might lead to unforeseen bugs in the future. You might want to consider only using one instance of `filePath`.

drustanyjt commented on src/main/java/demon/Demon.java:
```diff
@@ -0,0 +1,128 @@
+    public void run() {
+        Scanner sc = new Scanner(System.in);
+        ui = new Ui();
+        Commands command = new Commands(filePath);
```
@Colex2000 with reference to the above comment, here you are using the instance variable `this.filePath` (declared in line 17).

drustanyjt reviewed (commented):
I think there might be a potential bug in `src/main/java/demon/Demon.java` that you might want to pay attention to.

Colex2000 committed:
Bugfix: GUI chatbot response and window responsive

Colex2000 committed:
Merge branch 'Level-10'

Colex2000 committed:
Add assertions to validate internal assumptions

This commit introduces a series of assert statements across various methods to ensure that internal assumptions about the state and integrity of data are explicitly checked during development.

- Assert non-null task lists in task management methods to prevent NullPointerExceptions.

These assertions are intended for use in development and testing phases to catch logical errors and assumptions violations early. They are disabled in production to avoid impacting performance.

Colex2000 committed:
Merge pull request #1 from Colex2000/A-Assertions

Add assertions to validate internal assumptions

Colex2000 committed:
Refactor methods for improved abstraction

This commit restructures the methods in Command class to align with the Single Level of Abstraction Principle (SLAP). The main changes include:

- Abstracting all methods such that it is more readable.

Benefits Achieved:
- Enhanced readability as the methods has the same level of abstractions.
- Easier debugging as functionalities are isolated into their own methods.

Colex2000 committed:
Add duplicate task check functionality

This commit introduces new feature to check for duplicates when adding a new task. This feature aims to prevent the addition of tasks that are exactly identical in description and, where applicable, date and time for Deadline and Event tasks.

Key changes:
- Implemented `checkDuplicate` method to validate uniqueness across ToDo, Deadline, and Event tasks based on their descriptions and date/time.
- Refactored Deadline, Event, and Task class to accommodate the addition of the new feature.
- Introduced exception handling for duplicate tasks to notify users of attempted duplications and prevent task addition.

This functionality is crucial for maintaining an organized and efficient task list, ensuring that users do not inadvertently clutter their task list with duplicate entries.

Colex2000 committed:
Merge pull request #2 from Colex2000/A-CodeQuality

Refactor methods for improved abstraction

Colex2000 committed:
Merge branch 'master' into BCD-Extension

Colex2000 committed:
Update taskList.txt

Colex2000 committed:
Merge pull request #3 from Colex2000/BCD-Extension

Add duplicate task check functionality

Colex2000 committed:
Fix bug with the mainClass.set method

Colex2000 committed:
Merge branch 'master' of https://github.com/Colex2000/ip

Colex2000 committed:
Fix bug with the mainClass.set method

Colex2000 committed:
Refactor methods for improved abstraction

This commit restructures the methods to align with
the Single Level of Abstraction Principle (SLAP).
The main changes include:

- Abstracting most methods such that it is more readable.

Benefits Achieved:
- Enhanced readability.
- Easier debugging.

Colex2000 committed:
Add JUnit test to test addition of Todo task

Colex2000 committed:
Add JUnit test to test Command class add Deadline method

Colex2000 committed:
Bug fix expanding window dynamically

Colex2000 committed:
Add GUI window title

Colex2000 committed:
Fix bug uppercase description not captured

Colex2000 committed:
Add delay to exit to show exit message

Colex2000 committed:
Change UI message for unmark task

Colex2000 committed:
Add user guide for Demon chatbot

Add a comprehensive user guide to assist new users in navigating
through the Demon's features and functionalities.
The guide includes:

- Installation steps
- Basic usage and commands

This documentation aims to enhance the user experience by providing
clear instructions and support for our chatbot.

Colex2000 committed:
Add user guide for Demon chatbot

Add a comprehensive user guide to assist new users in navigating
through the Demon's features and functionalities.
The guide includes:

- Installation steps
- Basic usage and commands

This documentation aims to enhance the user experience by providing
clear instructions and support for our chatbot.

Colex2000 committed:
Fix bug to the gradle file

Colex2000 committed:
Remove JUnit test for more meaningful testing

Colex2000 committed:
Change string for filepath to relative path

Colex2000 committed:
Change file name for saving tasks

Colex2000 committed:
Add enable assertions when executing java code

Colex2000 committed:
Add directory for saving Demon.txt file

Colex2000 committed:
Fix bug image not showing

Colex2000 committed:
Revert "Fix bug image not showing"

This reverts commit b22a542d2e24a492efcfbb40a72f67e5b4261a9b.

Colex2000 committed:
Fix bug image not showing

Colex2000 committed:
Fix bug image not showing

Colex2000 committed:
Fix bug image not showing
